File: backend/app/models/solicitud_parametro.py
from datetime import datetime
from sqlalchemy import Column, Date, Integer, String, Text, TIMESTAMP, Boolean, JSON

from app.core.database import Base


class SolicitudParametro(Base):

    __tablename__ = "solicitud_parametro"

    oid = Column(Integer, primary_key=True, index=True)
    consecutivo = Column(String(50), nullable=True, index=True)
    tipo_parametro = Column(String(100), nullable=False)
    descripcion = Column(Text, nullable=False)
    fecha_apertura = Column(Date, nullable=True)
    fecha_cierre = Column(Date, nullable=True)
    hora_apertura = Column(String(20), nullable=True)
    hora_cierre = Column(String(20), nullable=True)
    total_valor = Column(Integer, nullable=True)
    total_unidad = Column(String(20), nullable=True)
    solicitante = Column(String(200), nullable=False)
    area = Column(String(200), nullable=True)
    ingreso = Column(String(50), nullable=True)
    medico = Column(String(200), nullable=True)
    tiempo_limite = Column(String(10), nullable=True)
    fecha_habilitacion = Column(TIMESTAMP, nullable=True)
    fecha_expiracion = Column(TIMESTAMP, nullable=True)
    estado = Column(String(50), nullable=False, default="Pendiente")
    motivo_rechazo = Column(Text, nullable=True)
    solicitud_extension = Column(String(100), nullable=True)
    observacion_resolucion = Column(Text, nullable=True)
    fecha_registro = Column(TIMESTAMP, nullable=False)


class ConfiguracionParametros(Base):
    __tablename__ = "configuracion_parametros"

    id = Column(Integer, primary_key=True, index=True)
    hc_default = Column(Integer, nullable=False, default=30)
    enf_hcrenf_default = Column(Integer, nullable=False, default=48)
    enf_haplmed_default = Column(Integer, nullable=False, default=48)
    hora_restablecimiento = Column(String(10), nullable=False, default="20:05")
    auto_restablecer = Column(Boolean, nullable=False, default=True)
    tipos_habilitados = Column(JSON, nullable=False, default=["Historia Clinica", "Enfermeria", "Otros"])
    tiempo_maximo_contador = Column(String(10), nullable=False, default="12:00")
    correos_historia_clinica = Column(Text, nullable=True)
    correos_enfermeria = Column(Text, nullable=True)
    correos_otros = Column(Text, nullable=True)
    updated_at = Column(TIMESTAMP, nullable=False, default=datetime.utcnow)


class ConfiguracionModulosInicio(Base):
    __tablename__ = "configuracion_modulos_inicio"

    id = Column(Integer, primary_key=True, index=True)
    modulos = Column(
        JSON,
        nullable=False,
        default={
            "coordinator": True,
            "creacionUsuario": True,
            "restablecimientoPassword": True,
            "validator": True,
            "solicitud": True,
        }
    )
    updated_at = Column(TIMESTAMP, nullable=False, default=datetime.utcnow)

File: frontend/src/types/solicitud-parametro.ts
export type EstadoSolicitud =
  | "Pendiente"
  | "Habilitado"
  | "Aprobado"
  | "Rechazado"
  | "Autorizado Solicitud Previa";

export interface SolicitudParametro {
  id: number;
  consecutivo: string;
  tipoParametro: "Enfermeria" | "Historia Clinica" | "Otros";
  descripcion: string;
  fechaApertura: string;
  fechaCierre: string;
  horaApertura: string;
  horaCierre: string;
  tiempoLimite?: string | null;
  totalValor: number | null;
  totalUnidad: string | null;
  solicitante: string;
  area: string;
  ingreso: string;
  medico: string;
  nombrePaciente: string;
  estado: EstadoSolicitud;
  motivoRechazo?: string | null;
  solicitudExtension?: string | null;
  observacionResolucion?: string | null;
  fechaRegistro: string;
  fechaHabilitacion?: string | null;
  fechaExpiracion?: string | null;
}

export interface ApiSolicitudParametro {
  oid: number;
  consecutivo: string | null;
  tipo_parametro: string;
  descripcion: string;
  fecha_apertura: string | null;
  fecha_cierre: string | null;
  hora_apertura: string | null;
  hora_cierre: string | null;
  tiempo_limite?: string | null;
  total_valor: number | null;
  total_unidad: string | null;
  solicitante: string;
  area: string | null;
  ingreso: string | null;
  medico: string | null;
  nombre_paciente: string | null;
  estado: string;
  motivo_rechazo: string | null;
  solicitud_extension: string | null;
  observacion_resolucion: string | null;
  fecha_registro: string;
  fecha_habilitacion?: string | null;
  fecha_expiracion?: string | null;
}

export interface ConfiguracionParametrosDTO {
  hc_default: number;
  enf_hcrenf_default: number;
  enf_haplmed_default: number;
  hora_restablecimiento: string;
  auto_restablecer: boolean;
  tipos_habilitados: string[];
  tiempo_maximo_contador?: string;
  correos_historia_clinica?: string;
  correos_enfermeria?: string;
  correos_otros?: string;
}

export interface ConfiguracionModulosInicioDTO {
  coordinator: boolean;
  creacionUsuario: boolean;
  restablecimientoPassword: boolean;
  validator: boolean;
  solicitud: boolean;
}

export const toSolicitudParametro = (item: ApiSolicitudParametro): SolicitudParametro => ({
  id: item.oid,
  consecutivo: item.consecutivo ?? `REQ-${item.oid}`,
  tipoParametro: item.tipo_parametro as SolicitudParametro["tipoParametro"],
  descripcion: item.descripcion,
  fechaApertura: item.fecha_apertura ?? "",
  fechaCierre: item.fecha_cierre ?? "",
  horaApertura: item.hora_apertura ?? "",
  horaCierre: item.hora_cierre ?? "",
  tiempoLimite: item.tiempo_limite,
  totalValor: item.total_valor,
  totalUnidad: item.total_unidad,
  solicitante: item.solicitante,
  area: item.area ?? "",
  ingreso: item.ingreso ?? "",
  medico: item.medico ?? "",
  nombrePaciente: item.nombre_paciente ?? "",
  estado: item.estado as EstadoSolicitud,
  motivoRechazo: item.motivo_rechazo,
  solicitudExtension: item.solicitud_extension,
  observacionResolucion: item.observacion_resolucion,
  fechaRegistro: item.fecha_registro,
  fechaHabilitacion: item.fecha_habilitacion,
  fechaExpiracion: item.fecha_expiracion,
});
